stdlib/src/builtin/string.mojo

```diff
@@ -609,6 +613,7 @@ struct String(
     fn __init__(inout self):
         """Construct an uninitialized string."""
         self._buffer = Self._buffer_type()
+        self._buffer.append(0)
```

Comment:
Note that this change means that constructing an empty string will now allocate, where it did not beforec. That may be what we want, but its a notable change to the default representation of a `String`.

E.g. a `List[String]` default initialized to empty `String` and then later filled in would now do O(N) allocations.

Reply:
> E.g. a List[String] default initialized to empty String and then later filled in would now do O(N) allocations.

I'm not sure I understand this part. Can you give a small code example?

Pratically speaking, when we have SSO, empty strings will have null termintors allocated on the stack, so it will be very cheap. I'd advise that until then, we err on the side of caution about safety in String.

Reply:
That's a compelling point regarding the small-string optimization.

I'm convinced now that NUL terminating by default is not necessarily a bad behavior to aim for long term🙂 

> Can you give a small code example? I'm not sure I see the regression you are talking about.

Not a regression per se, but consider:

```mojo
var list = List[String](String(), String(), String(), String(), String())

for count in len(list):
    list[count] = str(count)
```

Before this change, this program only allocates a `String` for each element once.

After this change, first five empty strings will be allocated, and then overwritten with the final content of the list as initialized by the `for` loop.

I don't necessarily think this is a convincing reason on its own not to move forward with this change. I'm just offering this as an example of the change in behavior this PR brings.

Reply:
Thank you for the insight. Indeed it's a good example where we may have degraded performance. With SSO, that shouldn't be an issue. I'd be interested to know if a null terminator needs to be a requirement for String, but I guess this can be discussed after SSO is done